### DISCUSSION TITLE
Invalid Mesh

### DISCUSSION THREAD

**Discussion by W1711** [visose/Robots#105] [Q&A] [answered]
Hello Vicente,

I was trying the robots plugin for some pick & place scripts. But every time I lose a mesh when I run the script. This varies between robot meshes and gripper meshes.

Example for the gripper, in its closed state, its mesh is correct but in its open state, the mesh is invalid.
For the robotic arm I use the Bartlett-UR10, and going through the script, the meshes of the arm axis get lost alternately.

This is due to what?

![0](https://user-images.githubusercontent.com/77907436/105619683-1955ee80-5dc3-11eb-89f9-583d0032c697.PNG)
![1](https://user-images.githubusercontent.com/77907436/105619684-19ee8500-5dc3-11eb-83a9-7cece9d13150.PNG)

**Comment by visose** [comment]
Hi, are you using the latest Rhino 7? Someone else contacted me with the exact same bug using Rhino 7. I can have a look if you attach a minimal rhino and grasshopper file that can reproduce the bug.
As a workaround you can manually orient the tool mesh to each target position using the Orient component and the resulting Planes for each target (from WorldXY plane to target 6th axis plane or from tool TCP plane to target TCP plane).

**Comment by W1711** [comment]
Yes I'm using Rhino 7 right now. I also think that it's about the Rhino version because I didn't have those kinds of problems with the Rhino6 version.
Here is a simple grasshopper script & the rhino file with the tools:

https://drive.google.com/file/d/15YFAN93U5syQl0xmOqi1c4ja58JMiwvS/view?usp=sharing
[MeshBug.zip](https://github.com/visose/Robots/files/5863386/MeshBug.zip)

Thank you!

**Comment by vinceHuyghe** [comment]
Have a look at the discussion in this thread it fixed the issue for me
https://discourse.mcneel.com/t/orient-mesh-creates-invalid-mesh/104082/2

**Comment by W1711** [comment]
It is true that rebuilding the mesh with Pufferfish can resolve the problem. For the tool it works fine, but for the robotic arm it would take a large amount of computational time, so it won't be efficient for a complex script (even for a simple one) :(

**Comment by visose** [comment, marked as answer]
Haven't had a look yet, but from the post above it looks like it's likely an issue with the original mesh. The mesh is not considered valid from the beginning, but this only gets picked up after transforming the mesh, and Grasshopper decides not to display meshes flagged as invalid. The solution would be to fix the original mesh. If it's a robot joint, by repairing the mesh in the .3dm of the robot library.

**Comment by W1711** [comment]
Solved, the meshes has to be fixed from the beginning. Thanks!

**Comment by UnfoldAntwerp** [comment]
@W1711 how did you fix this? I'm loading the UR10 from KULeuven and this one has the same issues. Looks like the other UR10s (Bartlett etc) are the same mesh.

**Comment by fabbas1** [comment]
a decent workaround is to instead of using a _custom preview_ node to simply activate preview on the _program simulation_ component depending on your application. gh/rhino doesnt seem to mind the invalid meshes in that case. works on the bartlett/leuven meshes in rhino7 for me.
![image](https://user-images.githubusercontent.com/66978550/151541080-5b487ecf-d59f-4783-8014-9ae6fc67f62e.png)

**Comment by visose** [comment]
To try resolve this issue, I wrote a script that goes through all the Rhino files of the libraries and attempts to fix the meshes. It does seem to have worked for the few that I tried but can't guarantee all of them.

It also deletes other unnecessary objects from the Rhino file. When installing these libraries through the new UI panel, the Rhino file is hidden from the user so any other objects in the file are unnecessary, better to keep the file more lightweight. You can still have auxiliary objects by keeping the library files in your Documents folder as before, this will override the online-installed library.

Below is a log of all the changes after running the script, I'll update them soon:

```
AA | Deleted empty layer (Default)
AA | Saved (AA.3dm)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Brep (Default)
Aarhus | Deleted Block definition (Block 01)
Aarhus | Deleted Block definition (Block 02)
Aarhus | Deleted Block definition (Block 03)
Aarhus | Deleted Block definition (Block 04)
Aarhus | Deleted empty layer (Default)
Aarhus | Saved (Aarhus.3dm)
AAU | Fixed Mesh (RobotArm.UR.UR10_4)
AAU | Fixed Mesh (RobotArm.UR.UR10_3)
AAU | Fixed Mesh (RobotArm.UR.UR10_2)
AAU | Fixed Mesh (RobotArm.UR.UR10_1)
AAU | Deleted empty layer (Layer 01)
AAU | Saved (AAU.3dm)
Automata | Fixed Mesh (RobotArm.Staubli.Eva_0)
Automata | Fixed Mesh (RobotArm.Staubli.Eva_5)
Automata | Saved (Automata.3dm)
Bartlett | Fixed Mesh (RobotArm.Staubli.TX200L_1)
Bartlett | Fixed Mesh (RobotArm.Staubli.TX200L_0)
Bartlett | Fixed Mesh (RobotArm.Staubli.TX200_1)
Bartlett | Fixed Mesh (RobotArm.Staubli.TX200_0)
Bartlett | Fixed Mesh (RobotArm.UR.UR10_1)
Bartlett | Fixed Mesh (RobotArm.UR.UR10_2)
Bartlett | Fixed Mesh (RobotArm.UR.UR10_3)
Bartlett | Fixed Mesh (RobotArm.UR.UR10_4)
Bartlett | Fixed Mesh (RobotArm.KUKA.KR210-2_1)
Bartlett | Deleted LineCurve (Positioner.KUKA.DKP-400)
Bartlett | Fixed Mesh (Positioner.KUKA.DKP-400_0)
Bartlett | Deleted ArcCurve (Positioner.KUKA.DKP-400)
Bartlett | Deleted empty layer (Default)
Bartlett | Saved (Bartlett.3dm)
Bigli | Fixed Mesh (RobotArm.KUKA.KR20-3_2)
Bigli | Deleted empty layer (Default)
Bigli | Saved (Bigli.3dm)
Carnegie | Fixed Mesh (RobotArm.ABB.IRB6640-235_5)
Carnegie | Fixed Mesh (RobotArm.ABB.IRB6640-235_1)
Carnegie | Saved (Carnegie.3dm)
CITA | Failed to fix Mesh (RobotArm.UR.UR5e_5)
CITA | Fixed Mesh (RobotArm.UR.UR5e_4)
CITA | Fixed Mesh (RobotArm.UR.UR5e_3)
CITA | Fixed Mesh (RobotArm.UR.UR5e_2)
CITA | Fixed Mesh (RobotArm.UR.UR5e_1)
CITA | Deleted empty layer (Default)
CITA | Saved (CITA.3dm)
Create | Fixed Mesh (Positioner.KUKA.KPF1-V500-V1_1)
Create | Fixed Mesh (Track.KUKA.KL4000_0)
Create | Failed to fix Mesh (Track.KUKA.KL4000_1)
Create | Fixed Mesh (RobotArm.KUKA.KR240_2)
Create | Failed to fix Mesh (RobotArm.KUKA.KR240_5)
Create | Fixed Mesh (RobotArm.KUKA.KR240_4)
Create | Deleted PolylineCurve (RobotArm.KUKA.KR240)
Create | Deleted Material (Custom)
Create | Deleted Material (Custom)
Create | Deleted Material (Custom)
Create | Deleted Material (Custom)
Create | Deleted Material (Custom)
Create | Deleted Material (Custom)
Create | Deleted Material (Metal)
Create | Deleted Material (Paint (1))
Create | Deleted Material (/Aluminum_Anodized_DarkGray (1))
Create | Deleted Material (/CarPaint_Simple_Black)
Create | Saved (Create.3dm)
CRR | Deleted PolylineCurve (Track.ABB.IRBT6004)
CRR | Fixed Mesh (Track.ABB.IRBT6004_1)
CRR | Fixed Mesh (RobotArm.ABB.IRB6700MH6-220_4)
CRR | Deleted Point (GuideData_Track)
CRR | Deleted LineCurve (GuideData_Track)
CRR | Deleted LineCurve (GuideData_Track)
CRR | Deleted Point (GuideData_Track)
CRR | Deleted empty layer (GuideData)
CRR | Deleted empty layer (Track)
CRR | Saved (CRR.3dm)
ETH | Deleted empty layer (Default)
ETH | Saved (ETH.3dm)
EX Architects | Fixed Mesh (RobotArm.ABB.IRB6700-155_5)
EX Architects | Deleted empty layer (Default)
EX Architects | Saved (EX Architects.3dm)
Grymsdyke | Fixed Mesh (RobotArm.KUKA.KR210-2_1)
Grymsdyke | Deleted empty layer (Default)
Grymsdyke | Saved (Grymsdyke.3dm)
IAAC | Deleted empty layer (Default)
IAAC | Saved (IAAC.3dm)
Innsbruck | Fixed Mesh (RobotArm.ABB.IRB2600-165_3)
Innsbruck | Fixed Mesh (RobotArm.ABB.IRB2600-165_4)
Innsbruck | Fixed Mesh (RobotArm.ABB.IRB2600-165_0)
Innsbruck | Deleted empty layer (Default)
Innsbruck | Saved (Innsbruck.3dm)
KU Leuven | Fixed Mesh (RobotArm.UR.UR5_0)
KU Leuven | Deleted Brep (RobotArm.UR.UR5)
KU Leuven | Fixed Mesh (RobotArm.UR.UR5_5)
KU Leuven | Fixed Mesh (RobotArm.UR.UR5_4)
KU Leuven | Fixed Mesh (RobotArm.UR.UR5_3)
KU Leuven | Fixed Mesh (RobotArm.UR.UR10_1)
KU Leuven | Fixed Mesh (RobotArm.UR.UR10_2)
KU Leuven | Fixed Mesh (RobotArm.UR.UR10_3)
KU Leuven | Fixed Mesh (RobotArm.UR.UR10_4)
KU Leuven | Saved (KU Leuven.3dm)
MIT Sim | Deleted Mesh in wrong layer (Default)
MIT Sim | Deleted Mesh in wrong layer (Default)
MIT Sim | Deleted Mesh in wrong layer (Default)
MIT Sim | Deleted Mesh in wrong layer (Default)
MIT Sim | Deleted Mesh in wrong layer (Default)
MIT Sim | Deleted Mesh in wrong layer (Default)
MIT Sim | Deleted Mesh in wrong layer (Default)
MIT Sim | Deleted Mesh in wrong layer (Default)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Mesh in wrong layer (RobotArm.ABB.IRB6640-280)
MIT Sim | Deleted Block definition (link_3)
MIT Sim | Deleted Block definition (link_1)
MIT Sim | Deleted Block definition (link_2)
MIT Sim | Deleted Block definition (link_3 01)
MIT Sim | Deleted Block definition (link_4)
MIT Sim | Deleted Block definition (link_6)
MIT Sim | Deleted Block definition (link_5)
MIT Sim | Deleted Block definition (piston)
MIT Sim | Deleted Block definition (cylinder)
MIT Sim | Deleted Block definition (piston 01)
MIT Sim | Deleted empty layer (Default)
MIT Sim | Saved (MIT Sim.3dm)
Nagami | Fixed Mesh (RobotArm.ABB.IRB6700-150_4)
Nagami | Fixed Mesh (RobotArm.ABB.IRB4600-205N_2)
Nagami | Deleted Mesh in wrong layer (RobotArm.ABB.IRB4600-205N_CopyFour)
Nagami | Fixed Mesh (RobotArm.ABB.IRB6640-235_4)
Nagami | Fixed Mesh (RobotArm.ABB.IRB6640-235_5)
Nagami | Fixed Mesh (RobotArm.ABB.IRB6640-235_3)
Nagami | Fixed Mesh (RobotArm.ABB.IRB6640-235_6)
Nagami | Deleted empty layer (CopyFour)
Nagami | Deleted empty layer (Layer 01)
Nagami | Deleted empty layer (Layer 02)
Nagami | Deleted empty layer (Layer 03)
Nagami | Deleted empty layer (Custom.ABB.MU100)
Nagami | Deleted empty layer (0)
Nagami | Deleted empty layer (1)
Nagami | Saved (Nagami.3dm)
Penn State | Deleted empty layer (Default)
Penn State | Saved (Penn State.3dm)
RMIT | Fixed Mesh (RobotArm.UR.UR10_1)
RMIT | Fixed Mesh (RobotArm.UR.UR10_2)
RMIT | Fixed Mesh (RobotArm.UR.UR10_3)
RMIT | Fixed Mesh (RobotArm.UR.UR10_4)
RMIT | Deleted empty layer (Default)
RMIT | Saved (RMIT.3dm)
Stuttgart | Fixed Mesh (RobotArm.ABB.IRB2600-165_3)
Stuttgart | Fixed Mesh (RobotArm.ABB.IRB2600-165_4)
Stuttgart | Fixed Mesh (RobotArm.ABB.IRB2600-165_0)
Stuttgart | Deleted empty layer (Default)
Stuttgart | Saved (Stuttgart.3dm)
Taiwan Tech | Deleted Mesh in wrong layer (Default)
Taiwan Tech | Fixed Mesh (Tool.KR60_AW14)
Taiwan Tech | Deleted ArcCurve (Positioner.KUKA.DKP-400)
Taiwan Tech | Fixed Mesh (Positioner.KUKA.DKP-400_0)
Taiwan Tech | Deleted LineCurve (Positioner.KUKA.DKP-400)
Taiwan Tech | Failed to fix Mesh (RobotArm.KUKA.KR6-Cell_1)
Taiwan Tech | Deleted empty layer (Default)
Taiwan Tech | Deleted empty layer (2)
Taiwan Tech | Deleted empty layer (3)
Taiwan Tech | Deleted empty layer (4)
Taiwan Tech | Deleted empty layer (5)
Taiwan Tech | Deleted empty layer (6)
Taiwan Tech | Saved (Taiwan Tech.3dm)
Taiwan | Fixed Mesh (RobotArm.KUKA.KR6-R700_3)
Taiwan | Fixed Mesh (RobotArm.KUKA.KR6-R700_1)
Taiwan | Fixed Mesh (RobotArm.KUKA.KR10-R1100_1)
Taiwan | Deleted empty layer (Default)
Taiwan | Saved (Taiwan.3dm)
UEL | Deleted Brep (RobotArm.ABB.IRB6620-220)
UEL | Deleted Brep (RobotArm.ABB.IRB6620-220)
UEL | Deleted Brep (RobotArm.ABB.IRB6620-220)
UEL | Deleted Brep (RobotArm.ABB.IRB6620-220)
UEL | Deleted InstanceReferenceGeometry (RobotArm.ABB.IRB6620-220)
UEL | Deleted Brep (RobotArm.ABB.IRB6620-220)
UEL | Deleted Brep (RobotArm.ABB.IRB6620-220)
UEL | Deleted Brep (RobotArm.ABB.IRB6620-220)
UEL | Deleted Brep (RobotArm.ABB.IRB6620-220)
UEL | Deleted Brep (RobotArm.ABB.IRB6620-220)
UEL | Deleted InstanceReferenceGeometry (RobotArm.ABB.IRB6620-220)
UEL | Deleted InstanceReferenceGeometry (RobotArm.ABB.IRB6620-220)
UEL | Deleted InstanceReferenceGeometry (RobotArm.ABB.IRB6620-220)
UEL | Deleted InstanceReferenceGeometry (RobotArm.ABB.IRB6620-220)
UEL | Deleted InstanceReferenceGeometry (RobotArm.ABB.IRB6620-220)
UEL | Deleted InstanceReferenceGeometry (RobotArm.ABB.IRB6620-220)
UEL | Deleted InstanceReferenceGeometry (RobotArm.ABB.IRB6620-220)
UEL | Deleted Block definition (IRB6620_220-150_m2004_rev0_01-6)
UEL | Deleted Block definition (Block 01)
UEL | Deleted Block definition (IRB6620_220-150_m2004_rev0_01-5)
UEL | Deleted Block definition (IRB6620_220-150_m2004_rev0_01-3)
UEL | Deleted Block definition (IRB6620_220-150_m2004_rev0_01-7)
UEL | Deleted Block definition (IRB6620_220-150_m2004_rev0_01-2)
UEL | Deleted Block definition (IRB6620_220-150_m2004_rev0_01-4)
UEL | Deleted Block definition (IRB6620_220-150_m2004_rev0_01-1)
UEL | Deleted Block definition (IRB6620_220-150_m2004_rev0_01)
UEL | Deleted empty layer (Default)
UEL | Saved (UEL.3dm)
UEM | Fixed Mesh (RobotArm.ABB.IRB2600-165_3)
UEM | Fixed Mesh (RobotArm.ABB.IRB2600-165_4)
UEM | Fixed Mesh (RobotArm.ABB.IRB2600-165_0)
UEM | Deleted empty layer (Default)
UEM | Saved (UEM.3dm)
UNIMELB | Deleted Brep (Default)
UNIMELB | Deleted Brep (Default)
UNIMELB | Deleted Brep (Default)
UNIMELB | Deleted Brep (Default)
UNIMELB | Deleted Brep (Default)
UNIMELB | Deleted Brep (Default)
UNIMELB | Deleted Brep (Default)
UNIMELB | Deleted Brep (Default)
UNIMELB | Deleted Brep (Default)
UNIMELB | Deleted Brep (Default)
UNIMELB | Deleted Brep (Default)
UNIMELB | Deleted Brep (Default)
UNIMELB | Deleted InstanceReferenceGeometry (Default)
UNIMELB | Fixed Mesh (RobotArm.ABB.IRB1200_3)
UNIMELB | Fixed Mesh (RobotArm.ABB.IRB1200_2)
UNIMELB | Deleted Block definition (Block 02)
UNIMELB | Deleted Block definition (swing)
UNIMELB | Deleted empty layer (Default)
UNIMELB | Saved (UNIMELB.3dm)
VUB | Deleted empty layer (Default)
VUB | Saved (VUB.3dm)
```

**Comment by v01tt** [comment]
I'm struggling with the same issue. Recently updated to Rhino 7. The files I prepared for the UR3e, seemed to work on R6.  Not so much on the R7. Would anyone be able to help take a look at the mesh for me? I've attached the required files here.
https://u.pcloud.link/publink/show?code=XZU127VZwBrsrEtQqmFdh1CWRXKkN7IwvNqk

**Reply by visose** [reply]
If you select each mesh and run the _what command, you'll see that some of them will show "invalid mesh". Run the _meshrepair command on them and fix zero length edges, degenerate faces, duplicate faces, and inconsistent normals. Having non-manifold edges and disjoint pieces is fine, no need to repair these.

Also, the file size is very large. Run the _reducemesh command to reduce the mesh sizes by 90% or so. You can use the Rendered view and check before/after (undo/redo) to see how much the reduce mesh has affected quality, you will notice its not that much. You can do this selectively, like keep the 6th joint with higher resolution, and lower in other.

Check again for invalid meshes after running _reducemesh.

**Reply by visose** [reply]
Also, use the _blockmanager command to delete all the block definitions, or use the _purge command. These make up half the file size.

**Reply by v01tt** [reply]
Thank you Visose, for not only addressing concerns, but taking the time to walk us through the fixes. 
Followed the above steps. Worked perfectly. Reduced file size from 75mb down to 7mb. And all meshes are Valid.

**Comment by Braghebrevi** [comment]
Hello, I have a very similar problem. 
Meshes are rendered for an instant and then disappear. As long as the canvas is not touched and de view(port) does not change, the rendered geometry persists, but at the smallest movement it disappears. Manually changing the "program simulation" time resets the preview, but once again any movement on the canvas or viewport makes them disappear. 
Any clue what could this be?

**Reply by Braghebrevi** [reply]
NB: meshes have been thoroughly checked, both the robot themselves and the effector tool, but they are all valid.

**Reply by visose** [reply]
Are you using any of the existing libraries, or did you generate your own robot meshes?
Could you attach a minimal reproducible example?

**Reply by Braghebrevi** [reply]
Thank you for the quick reply!

First pic, the moment I connect the "program simulation" to the program output, second pic a moment after, when i click on the canvas. Preview mesh just disappears. 

<img width="1728" alt="Screenshot 2022-09-20 at 23 55 08" src="https://user-images.githubusercontent.com/20628253/191371946-494f9bcd-3975-4e03-9ffc-e61bc8bd9ee1.png">

<img width="1728" alt="Screenshot 2022-09-20 at 23 55 17" src="https://user-images.githubusercontent.com/20628253/191372330-8f8bc398-c106-4b82-aa92-675f5b65d4e2.png">

In the pic you can see all the meshes being valid. There are four programs being input into the simulation at the moment so not all meshes in the panel can be seen, but they are all valid. I have tested with single program only and result is the same. 
Also, always in this pic the program(s) are output from a cluster into the "program simulation" but same happens if all is outside cluster. 

Two more factors to be considered: a) everything worked perfectly fine until two days ago (nothing change from the robotics point of view, only minimally the target trajectories) and b) I found that disabling the custom tool effector created with the "create tool" component solves the issue (but really doesn't because hey, no tool). Yes tool meshes are valid and were already simplified/checked before this reply, with no effect)

Here other four pics: first robots being rendered for a brief moment after disabling/re-enabling the simulation; second the robot disappearing as soon  as any interaction with canvas/viewport happens, and third robots re-appearing but this time permanently, if custom tool is disabled. Fourth is the content of the custom tool cluster.

<img width="1728" alt="Screenshot 2022-09-21 at 00 08 11" src="https://user-images.githubusercontent.com/20628253/191373724-73e62ac6-ba65-4e66-ba21-0b3c039a6509.png">

<img width="1728" alt="Screenshot 2022-09-21 at 00 08 20" src="https://user-images.githubusercontent.com/20628253/191373961-7f4c76ba-1c87-4111-b17a-09b9be959c26.png">

<img width="1728" alt="Screenshot 2022-09-21 at 00 08 29" src="https://user-images.githubusercontent.com/20628253/191373977-628364bf-a1ba-43bf-81cf-d7ed302efcf0.png">

<img width="1728" alt="Screenshot 2022-09-21 at 00 16 47" src="https://user-images.githubusercontent.com/20628253/191374666-6f0db4b2-4fc9-4a3f-86cb-b8c8f00c90f3.png">

**Reply by visose** [reply]
Sorry, I meant if you could attach the files required to be able to reproduce this issue in another computer.
This would be one grasshopper file, maybe a Rhino file if strictly necessary, and if you are using a custom robot library, the .3dm/.xml pair.

Before attaching please remove all components that are unrelated to the issue. See how far you can go simplifying the logic of your file while still reproducing the issue. Don't include components from third party plugins besides Robots.

**Reply by Braghebrevi** [reply]
Sorry 😅
Here it is. I inserted scribbles to define/comment the various parts.
I think it is as simple as possible: just defining a random trajectory, altering it a little, defining the tool, defining the targets, all into program, all into simulation.

https://file.io/CUPG9FS9zwAd

Note that as written beside the tool geometry, by disabling it the preview works despite the invalid mesh that is created.
Instead, the perfectly normal mesh breaks all.

An observation: I have noticed that when the tool mesh is active, before interaction that make everything disappear occur, the preview aspect of the robot+tool assembly changes slightly, going more transparent than the default one. I think it may be related to how the components handles the additional mesh, but this is just an hypotesis. You can test this behaviour by rapidly enabling/disabling the mesh of the custom tool.

Here a before/after pic. 
A) custom mesh disabled, darker, stable preview
B) custom mesh active, lighter, preview disappears after interaction

<img width="942" alt="Screenshot 2022-09-21 at 10 43 55" src="https://user-images.githubusercontent.com/20628253/191458976-55e6a8d7-88c0-4e4c-b1e3-9e990e521c99.png">

<img width="943" alt="Screenshot 2022-09-21 at 10 44 07" src="https://user-images.githubusercontent.com/20628253/191458911-7a9b1576-288d-49d0-ae6b-a2b031c9c3fc.png">



Final note: in my gh definition which is much larger and where the Robots plugin components reside inside a cluster, the preview is not permanent at all. If they appear at all after refresh, they instantly blink out.

Thanks again!

**Reply by visose** [reply]
Hi, I can't open the file as it contains components from a third-party plugin that I don't have installed.
If the issue only appears with these components, maybe it's caused by the third-party plugin.

Also, are you sure you can't simplify further the definition while retaining the issue?
For example, can you reproduce the issue without changing the robot origin, not defining a tool and creating a program with only two targets?
If not, what's the minimum number of targets needed to reproduce the issue? What's the simplest tool that reproduces the issue? There's a group of components called "visualization of targets", I guess those are not related to reproducing the issue?

**Reply by Braghebrevi** [reply]
Sorry, i had a parakeet component I forgot to remove. 

Here the previous file with detailed information cleaned of other plugins  +  the simplest possible replication of the error: it is related to the create tool component, which I am required to use - https://file.io/dkYU5BCpSF34

**Reply by visose** [reply]
I don't think I can reproduce the issue. All meshes show as valid, and preview properly after animating the slider and moving the viewport camera. If I disable the tool mesh, only the tool mesh becomes invalid, but this is expected as it's creating a mesh with 0 vertices, which is invalid, and this doesn't affect the other meshes.

<img width="1231" alt="image" src="https://user-images.githubusercontent.com/5129614/191555448-d2a08b83-70dd-41a5-b3b0-e4e50e2e9eee.png">

When I opened your file, it was missing a reference to the point parameter, so I added a panel with some coordinates. I also added a slider and enabled the mesh parameter.

Do you have any other clues on how I could reproduce this issue?
Do you have access to another computer where you can test this?
Does this only happen with the Robots plugin?
Could it be a graphics card issue?

Make sure you have the latest Rhino 7 service release and the latest Robots version installed.

**Reply by Braghebrevi** [reply]
The tcp parameter was missing, sorry.

Anyway, i have the latest version of Rhino, Robots, and Robots extensions.
This does happen only on the robots plugin, exclusively when using the create tool component.
It could be a graphic card issue (working on a M1 Max Mac book pro, which i know is a perky system) but i don't understand how this didn't happen since 3 days ago, while until then it performed flawlessly.
I also completely reset gh display settings in the remote case it was something related to that.

I just tried on my old mac book pro and everything works. I now have no idea how to address it on my current system.

**Reply by visose** [reply]
I understand that you could not reproduce the issue on another computer, and is likely a graphics card or driver issue?
